### ISSUE TITLE
Extend `DatabaseValueConvertible` JSON parsing format customization

### ISSUE THREAD

**Issue by JhonnyBillM** [groue/GRDB.swift#1428]
### What did you do?

Added a `Codable` struct with `Double` values that can be `.infinity` as a field in a table. As expected, this throws an error: 

> Unable to encode Double.infinity directly in JSON. Use JSONEncoder.NonConformingFloatEncodingStrategy.convertToString to specify how the value should be encoded.

### What did you expect to happen?

That we could customize the `JSONDecoder` and `JSONEncoder` just as we can do for `FetchableRecord` and `EncodableRecord`

```swift
extension DataWeNeedButNotNormalized: DatabaseValueConvertible {
    public static func databaseJSONDecoder() -> JSONDecoder {
        customDecoder
    }

    public static func databaseJSONEncoder() -> JSONEncoder {
        customEncoder
    }
}
```

### What happened instead?

Because `DatabaseValueConvertible` doesn't have methods to customize the `JSONDecoder` and `JSONEncoder` used, we had to manually conform to the protocol, which it is easy, but as a convenience we think it would be nice to match `FetchableRecord` and `EncodableRecord` here and allow `DatabaseValueConvertible` to still be auto-synthesized for `Codable` entities while providing a custom Decoder and/or Encoder.

**In summary**, I suggest we add the following two methods to `DatabaseValueConvertible`:

`public static func databaseJSONDecoder() -> JSONDecoder`
`public static func databaseJSONEncoder() -> JSONEncoder`

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 6.10.1
**Installation method:** CocoaPods
**Xcode version:** 14.3.1
**Swift version:** 5.6
**Platform(s) running GRDB:** iOS, watchOS and tvOS
**macOS version running Xcode:** 13.4

### Demo Project

[GRDBAsyncDemo+InfinityValueTest.zip](https://github.com/groue/GRDB.swift/files/12571449/GRDBAsyncDemo%2BInfinityValueTest.zip)

**Comment by JhonnyBillM** [comment]
Opened #1429, but I understand that in the bigger scope of things, this is a very small and targeted issue, so I would totally understand if you feel like this protocol should not be expanded, or any other comments or suggestions you may have.

Thank you in advance and thanks for all your work on this great library.

**Comment by JhonnyBillM** [comment]
Great point, as some initial friction in that implementation was what gave me the initial impression of "this should be consistent with FetchableRecord". Without the new APIs, it would look like this:
```swift
private let customDecoder: JSONDecoder = {
    let decoder = JSONDecoder()
    decoder.dataDecodingStrategy = .base64
    decoder.dateDecodingStrategy = .millisecondsSince1970
    decoder.nonConformingFloatDecodingStrategy = .convertFromString(
        positiveInfinity: "+InF",
        negativeInfinity: "-InF",
        nan: "NaN"
    )
    return decoder
}()

private let customEncoder: JSONEncoder = {
    let encoder = JSONEncoder()
    encoder.dataEncodingStrategy = .base64
    encoder.dateEncodingStrategy = .millisecondsSince1970
    encoder.nonConformingFloatEncodingStrategy = .convertToString(
        positiveInfinity: "+InF",
        negativeInfinity: "-InF",
        nan: "NaN"
    )
    // guarantee some stability in order to ease record comparison
    encoder.outputFormatting = .sortedKeys
    return encoder
}()

extension MyValue: DatabaseValueConvertible {
    var databaseValue: DatabaseValue {
        let data = try? customEncoder.encode(self)
        guard let value = DatabaseValue(value: data as Any) else {
            return .null
        }
        return value
    }

    static func fromDatabaseValue(_ dbValue: DatabaseValue) -> Self? {
        switch dbValue.storage {
        case .null, .int64, .double, .string:
            return nil
        case let .blob(data):
            return try? customDecoder.decode(Self.self, from: data)
        }
    }
}
```

**Comment by groue** [comment]
All right! Thank you @JhonnyBillM, this is a great sample code. We're going to have #1429 merged :-)

Now, the proper implementation of `DatabaseValueConvertible` should read instead:

```swift
extension MyValue: DatabaseValueConvertible {
    var databaseValue: DatabaseValue {
        // Until GRDB make it possible to handle errors here,
        // It's better to crash than to store NULL and lose data.
        // Silently losing data is the ABSOLUTE NO-NO in this library.
        //
        // This is also consistent with the current semantics of the
        // DatabaseValueConvertible protocol: it is reserved to
        // types that CAN be stored in the database.
        try! customEncoder.encode(self).databaseValue
    }

    static func fromDatabaseValue(_ dbValue: DatabaseValue) -> Self? {
        // It's important to support both BLOB and TEXT database values
        // when decoding JSON. BLOB is OK, and some people use it.
        // Yet SQLite strongly fosters TEXT:
        // <https://www.sqlite.org/json1.html#interface_overview>
        guard let data = Data.fromDatabaseValue(dbValue) else {
            // Not BLOB, not TEXT
            return nil
        }
        return try? customDecoder.decode(Self.self, from: data)
    }
}
```

After all, it is not as "easy" as one could expect 😅 This means that there is a net advantage if the library takes care of it, and this is why #1429 is a great pull request 👍 

Maybe the PR will have to be slightly revisited in light of the two rules exposed in the above sample code (no silent data loss, and support for both BLOB and TEXT)?

**Comment by JhonnyBillM** [comment]
Thank you for this snippet and the comments. Really helpful !
As for the PR, it is already following your snippet, as it is doing what the library was already doing; just replaced the encoder/decoder with the protocol methods. 
Now, the part that I don't love is having to introduce a `jsonEncoder` parameter and [perform this safeguard](https://github.com/JhonnyBillM/GRDB.swift/blob/eec476b9ba6cfad529a6fb0592b4d1be7b339aa4/GRDB/Core/Support/StandardLibrary/DatabaseValueConvertible%2BEncodable.swift#L123-L132) in `DatabaseValueEncoder`. We could discuss it in the PR if you think we should handle this differently

**Comment by groue** [comment]
> As for the PR, it is already following your snippet, as it is doing what the library was already doing; just replaced the encoder/decoder with the protocol methods.

Yes, this is perfect 👍 

> Now, the part that I don't love is having to introduce a `jsonEncoder` parameter and [perform this safeguard](https://github.com/JhonnyBillM/GRDB.swift/blob/eec476b9ba6cfad529a6fb0592b4d1be7b339aa4/GRDB/Core/Support/StandardLibrary/DatabaseValueConvertible%2BEncodable.swift#L123-L132) in `DatabaseValueEncoder`. We could discuss it in the PR if you think we should handle this differently

I fully agree. I added 746f65a9793ec3bce16b440e9a642205b6881823 to fix it. Handling of Codable is not trivial, and this area of the GRDB code is not the easiest to follow. Actually I just made the jsonEncoder not optional (because we *want* the user to customize it), and the rest has followed.

**Comment by groue** [comment]
Closing in favor of #1429
